### ISSUE TITLE
Callback on browser request

### ISSUE THREAD

**Issue by lorenzoh** [tlienart/LiveServer.jl#119]
I have a use case for triggering a callback on a request *before* the file is returned for my documentation package [Pollen.jl](https://github.com/lorenzoh/Pollen.jl). It has a *lazy* mode where, instead of building all the pages first, it only builds a page when it is requested in the browser (saving tons of time when doing interactive work). To use it with LiveServer.jl requires executing a callback before the request returns.

Implementation-wise this boils down to adding a  `requestcb` keyword argument to `LiverServer.serve` and calling it in the handler like

```julia
function serve(fw::FileWatcher=SimpleWatcher(file_changed_callback);
               requestcb = req -> (),
               host::String="127.0.0.1", port::Int=8000, dir::AbstractString="", verbose::Bool=false,
               coreloopfun::Function=(c,fw)->nothing,
               inject_browser_reload_script::Bool = true,
               launch_browser::Bool = false)

    ...

    # make request handler
    req_handler = HTTP.RequestHandlerFunction() do req
        callback(req)
        serve_file(fw, req; inject_browser_reload_script = inject_browser_reload_script)
    end
```

If this seems like a thing you'd be okay with adding, I'll open up a PR with the implementation.

**Comment by tlienart** [comment]
I'm ok with this with just a couple of very minor notes

1. please consider calling this `preprocess_request` so that it's more explicit (or anything of the kind that is more explicit about the role of the function)
2. please copy paste a version of what you wrote above in the docstring to clarify what that could be used for 
3. please add a type (I know, I know)

Basically what you want is that if you change ABC.md but the browser is not at the corresponding page, no triggering happens? if so, fine, I'll need to try your package to see why that's useful but I'll happily admit that you've thought more about this than I have.

**Comment by tlienart** [comment]
PS: out of curiosity, do you have a demo repo that shows how Pollen works?

**Comment by lorenzoh** [comment]
> Basically what you want is that if you change ABC.md but the browser is not at the corresponding page, no triggering happens?

If I request a page that hasn't been built yet, I want to trigger the build before the page returns. Since a lot of the documentation pages have lots of code examples, waiting for them all to run takes a long time. This way only the pages that are requested are built.

> please consider calling this preprocess_request so that it's more explicit (or anything of the kind that is more explicit about the role of the function)

In my use case, nothing happens to the request itself, but I guess it could be useful if a new request can be returned from the handler. 

> please add a type (I know, I know)

Do you mean `::Function`? There isn't a good way to restrict the type of an argument to something callable; using `Function` would mean you couldn't pass a `struct` with an overloaded function call.

> PS: out of curiosity, do you have a demo repo that shows how Pollen works?

It is not released yet, but I am already using it for the following documentation pages:
- https://fluxml.ai/FastAI.jl/dev/README.md.html
- https://lorenzoh.github.io/DataAugmentation.jl/dev/README.md.html

**Comment by tlienart** [comment]
> Do you mean `::Function`? There isn't a good way to restrict the type of an argument to something callable; using `Function` would mean you couldn't pass a `struct` with an overloaded function call.

yes, is the latter an issue for you? if so leave as you suggested, it's fine, just make the default something that Julia can consider a no-op like `identity`.

Cool I'll check out the rest!

**Comment by tlienart** [comment]
Ah last question, did Franklin (with a custom template based on hugobooks) not suit your purpose?

**Comment by lorenzoh** [comment]
Pollen.jl was mostly inspired by Publish.jl and solves some pain points of it. A big difference is that Pollen.jl is not a static site generator, but a document generation and transformation framework. Pollen projects consist of collections of virtual documents that can be transformed by pluggable rewriters. I am currently using it to generate documentation, but it is flexible enough for many other things. For example, I am thinking about using it as a literate package development tool similar to [nbdev](https://nbdev.fast.ai/).

The way it's written allows for composing different transformation steps naturally, since they are all operating on virtual documents which are represented as tree-like data structures with many rewriting transformations available. You can see [here](https://github.com/lorenzoh/Pollen.jl/blob/main/src/projects.jl) how different rewriting steps are composed to get a project for documentation.

Additionally, it makes sure that it's possible to incrementally rebuild any document or part of a document; when you change a literate code cell only that cell should be rerun and the document rebuilt, not the whole project.

From what I've seen, all of these things would have been hard to integrate into Publish.jl or Franklin.jl

**Comment by tlienart** [comment]
Thanks for the explanation, it makes sense to me! good luck :)

**Comment by tlienart** [comment]
closed by #123
